### PR TITLE
Add rector

### PR DESCRIPTION
**Pull request by VincentLanglet**
https://github.com/sonata-project/dev-kit/pull/2106

### CODE REVIEW

**Comment by jordisala1991** [comment]
I dont think it is possible to do on 3.x without breaking everything. It needs to be done on 4.x

**Comment by VincentLanglet** [comment]
> I dont think it is possible to do on 3.x without breaking everything. It needs to be done on 4.x

Why would it break something ? The change made are supposed to be BC.
Do you have any BC-break in mind ?

But, I prefer to finish this PR after we solved https://github.com/sonata-project/SonataPageBundle/issues/1442 to avoid having wrong typehint.

**Comment by jordisala1991** [comment]
Adding typehints to all properties is not Bc break?

**Comment by VincentLanglet** [comment]
> Adding typehints to all properties is not Bc break?

I had no error for
```
class A
{
     protected A $foo;
}

class B extends A
{
	protected $foo;
}
```
But trying with another sandbox there is an error indeed. ^^'

I'll try to disable some rules.
We could add a check for page-bundle 3.x in https://github.com/sonata-project/dev-kit/blob/master/templates/project/rector.php.twig#L37

This would 
- allow to have rector on 4.x
- allow to remove the rector config key in dev-kit.

**Comment by VincentLanglet** [comment]
Rector is not supposed to add BC break. I found the issue, it's the `@final` annotation.
I created an issue https://github.com/rectorphp/rector/issues/7289

**Comment by eerison** [comment]
> > Adding typehints to all properties is not Bc break?
> 
> I had no error for
> 
> ```
> class A
> {
>      protected A $foo;
> }
> 
> class B extends A
> {
> 	protected $foo;
> }
> ```
> 
> But trying with another sandbox there is an error indeed. ^^'
> 
> I'll try to disable some rules. We could add a check for page-bundle 3.x in https://github.com/sonata-project/dev-kit/blob/master/templates/project/rector.php.twig#L37
> 
> This would
> 
> * allow to have rector on 4.x
> * allow to remove the rector config key in dev-kit.

What do you think to use `LevelSetList::UP_TO_PHP_73` and enable 7.4 only in 4.x branch?

**Comment by VincentLanglet** [comment]
> What do you think to use LevelSetList::UP_TO_PHP_73 and enable 7.4 only in 4.x branch?

Dev kit automatically use the lower php version of the branch (which is now 7.4)
So anyway, it require some extra dev on dev-kit

**Comment by VincentLanglet** [comment]
Now https://github.com/rectorphp/rector/issues/7289 is merged, this is BC @jordisala1991 
But I have to target dev-main atm.

File: src/Request/RequestFactory.php [hunk 1]
```diff
@@ -66,17 +66,7 @@ private static function configureFactory($type)
...
-            array $request = [],
-            array $attributes = [],
-            array $cookies = [],
-            array $files = [],
-            array $server = [],
-            $content = null
-        ) {
-            return new SiteRequest($query, $request, $attributes, $cookies, $files, $server, $content);
-        });
+        Request::setFactory(static fn (array $query = [], array $request = [], array $attributes = [], array $cookies = [], array $files = [], array $server = [], $content = null) => new SiteRequest($query, $request, $attributes, $cookies, $files, $server, $content));
```
Comment (jordisala1991) [inline src/Request/RequestFactory.php]: This would be nice to do some line breaks.

File: src/Request/SiteRequestContext.php [hunk 2]
```diff
@@ -31,10 +31,7 @@ class SiteRequestContext extends RequestContext implements SiteRequestContextInt
      */
     protected $selector;
 
-    /**
-     * @var SiteInterface
-     */
-    private $site;
+    private ?SiteInterface $site = null;
```
Comment (jordisala1991) [inline src/Request/SiteRequestContext.php:34]: is it nullable?

File: src/Twig/Extension/PageExtension.php [hunk 3]
```diff
@@ -39,45 +39,24 @@
...
     /**
      * @var array
      */
     private $resources;
 
-    /**
-     * @var Environment
-     */
-    private $environment;
+    private ?Environment $environment = null;
```
Comment (jordisala1991) [inline src/Twig/Extension/PageExtension.php:51]: is it nullable?

File: tests/Entity/Snapshot.php [hunk 4]
```diff
@@ -17,10 +17,7 @@
 
 final class Snapshot extends BaseSnapshot
 {
-    /**
-     * @var int
-     */
-    protected $id;
+    protected ?int $id = null;
```
Comment (jordisala1991) [inline tests/Entity/Snapshot.php:20]: is it nullable?

File: tests/Site/HostPathSiteSelectorTest.php [hunk 5]
```diff
@@ -155,9 +155,7 @@ final class HostPathSiteSelector extends BaseSiteSelector
      */
     public static function _camelize($property): string
     {
-        return preg_replace_callback('/(^|[_. ])+(.)/', static function ($match) {
-            return ('.' === $match[1] ? '_' : '').strtoupper($match[2]);
-        }, $property);
+        return preg_replace_callback('/(^|[_. ])+(.)/', static fn ($match) => ('.' === $match[1] ? '_' : '').strtoupper($match[2]), $property);
```
Comment (jordisala1991) [inline tests/Site/HostPathSiteSelectorTest.php]: I would break this in one line per argument.

File: tests/Site/HostSiteSelectorTest.php [hunk 6]
```diff
@@ -115,9 +115,7 @@ final class HostSiteSelector extends BaseSiteSelector
      */
     public static function _camelize($property): string
     {
-        return preg_replace_callback('/(^|[_. ])+(.)/', static function ($match) {
-            return ('.' === $match[1] ? '_' : '').strtoupper($match[2]);
-        }, $property);
+        return preg_replace_callback('/(^|[_. ])+(.)/', static fn ($match) => ('.' === $match[1] ? '_' : '').strtoupper($match[2]), $property);
```
Comment (jordisala1991) [inline tests/Site/HostSiteSelectorTest.php]: same as before

File: src/Model/SnapshotPageProxy.php [hunk 7]
```diff
@@ -45,12 +36,9 @@ class SnapshotPageProxy implements SnapshotPageProxyInterface
     /**
      * @var PageInterface[]
      */
-    private $parents;
+    private ?array $parents = null;
```
Comment (jordisala1991) [inline src/Model/SnapshotPageProxy.php:43]: is it nullable?

File: src/Model/SnapshotPageProxy.php [hunk 8]
```diff
@@ -22,20 +22,11 @@
...
-     * @var SnapshotInterface
-     */
-    private $snapshot;
+    private SnapshotInterface $snapshot;
 
-    /**
-     * @var PageInterface
-     */
-    private $page;
+    private ?PageInterface $page = null;
```
Comment (jordisala1991) [inline src/Model/SnapshotPageProxy.php:29]: is it nullable?

**Review by jordisala1991 — COMMENTED** [review]
Nice, it will be hard to merge to 4.x tho

File: src/Request/SiteRequestContext.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Request/SiteRequestContext.php:34]: It's not initiliazed.

```
final public function setSite(SiteInterface $site)
    {
        $this->site = $site;
    }

    final public function getSite()
    {
        if (!$this->site instanceof SiteInterface) {
            $this->site = $this->selector->retrieve();
        }

        return $this->site;
    }
```
Coming from a setter.

File: src/Twig/Extension/PageExtension.php (hunk 3 above)
Comment (VincentLanglet) [reply src/Twig/Extension/PageExtension.php:51]: Not initialized in the constructor but in
```
public function initRuntime(Environment $environment)
    {
        $this->environment = $environment;
    }
```

File: src/Model/SnapshotPageProxy.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Model/SnapshotPageProxy.php:43]: Yes when it's not initialized 
```
public function getParents()
    {
        if (!$this->parents) {
```

File: src/Model/SnapshotPageProxy.php (hunk 8 above)
Comment (VincentLanglet) [reply src/Model/SnapshotPageProxy.php:29]: Yes when it's not loaded
```
private function load()
    {
        if (!$this->page && $this->transformer) {
            $this->page = $this->transformer->load($this->snapshot);
        }
    }
```

File: tests/Entity/Snapshot.php (hunk 4 above)
Comment (VincentLanglet) [reply tests/Entity/Snapshot.php:20]: It's not initialized.

**Review by jordisala1991 — APPROVED** [review]

File: composer.json [hunk 9]
```diff
@@ -65,6 +65,7 @@
         "phpunit/phpunit": "^9.5",
         "psalm/plugin-phpunit": "^0.16",
         "psalm/plugin-symfony": "^3.0",
+        "rector/rector": "dev-main",
```
Comment (jordisala1991) [inline composer.json:68]: I think rector just released a patch version. Can you see if the fix is included?

**Comment by VincentLanglet** [comment]
It is invluded

**Pull request merged**